### DISCUSSION TITLE
Custom Actions with .Net 7

### DISCUSSION THREAD

**Discussion by mjanulaitis** [wixtoolset/issues#7136] [Questions]
I am trying to create a custom action using .Net 7.  I don't see any documentation for .Net Core custom actions.  I stumbled upon WixToolset.Dtf.WindowsInstaller in NuGet and was able to get the [CustomAction] attribute to be recognized however when I build I get the following error:

EXEC : error : Invalid argument: WixToolset.Dtf.WindowsInstaller.dll must be included in the list of support files. If using the MSBuild targets, make sure the assembly reference has the Private (Copy Local) flag set.

Is this the correct package to reference?  If so, what changes do I need to make to get this to work?

**Comment by barnson** [comment]
DTF only supports .NET Framework-based custom actions (for now).

**Reply by mjanulaitis** [reply]
@barnson OK thanks.  I have updated my new project to use .Net 4.7.2.

**Reply by nickhoeferpickpro** [reply]
then how do we run custom actions for .net core?

**Reply by HungryDoctor** [reply]
@mjanulaitis, there is a WA for the issue. You can use `<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>`. But the real issue is that `WixToolset.Dtf.MakeSfxCA.exe` is built to net472 and it won't be able to load net7.0 libraries. Even if you put full .net7.0 runtime to the exe file.
I tried to build `MakeSfxCA` to net6.0, but since there are assembly loading changes (`Assembly.ReflectionOnlyLoad` is deprecated), it won't work so simple. If `Assembly.ReflectionOnlyLoad` is replaced with `Assembly.Load` it still won't work out of the box because of .net assembly loading mechanism is different.
@barnson, is there any workaround for targeting .net6.0 custom actions? If there is no WA, will this feature be available in v5?

**Reply by HungryDoctor** [reply]
@barnson, are there any possible issues besides reflection? If no, I can try to rewrite the reflection loading part with [MetadataLoadContext](https://github.com/dotnet/runtime/issues/15033).

**Reply by barnson** [reply]
It requires hosting .NET Core, which is very different than hosting .NET Framework.

**Reply by HungryDoctor** [reply]
I see. Hopefully, the feature will be available in the next releases

**Reply by barnson** [reply]
Only if someone implements it.
